apps/src/blockly/customBlocks/googleBlockly/behaviorBlocks.js

```diff
@@ -0,0 +1,256 @@
+);
+
+// This extension adds an SVG frame around behavior definition blocks.
+// Not used when the modal function is enabled.
+GoogleBlockly.Extensions.register('behaviors_block_frame', function () {
+  if (!useModalFunctionEditor && !this.workspace.noFunctionBlockFrame) {
+    this.functionalSvg_ = new BlockSvgFrame(
+      this,
+      msg.behaviorEditorHeader(),
+      'blocklyFunctionalFrame'
+    );
+
+    this.setOnChange(function () {
+      if (!this.isInFlyout) {
+        this.functionalSvg_.render(this.svgGroup_, this.RTL);
+      }
+    });
+  }
+});
+
+GoogleBlockly.Extensions.registerMutator(
+  'behavior_get_mutator',
+  behaviorGetMutator
+);
+
+/**
+ * Constructs the blocks required by the flyout for the procedure category.
+ * Modeled after core Blockly procedures flyout category, but excludes unwanted blocks.
+ * Derived from core Google Blockly:
+ * https://github.com/google/blockly/blob/5a23c84e6ef9c0b2bbd503ad9f58fa86db1232a8/core/procedures.ts#L202-L287
+ * @param {WorkspaceSvg} workspace The workspace containing procedures.
+ * @returns an array of XML block elements
+ */
+export function flyoutCategory(workspace) {
+  const blockList = [];
+
+  const newBehaviorButton = {
+    kind: 'button',
+    text: msg.createBlocklyBehavior(),
+    callbackKey: 'createNewBehavior',
+  };
+
+  const behaviorDefinitionBlock = {
+    kind: 'block',
+    type: 'behavior_definition',
+    fields: {
+      NAME: Blockly.Msg.PROCEDURES_DEFNORETURN_PROCEDURE,
+    },
+  };
+
+  const createNewBehavior = function () {
+    // Everything here is place-holder code that should be replaced with a
+    // call to open the behavior editor with a new defintion block.
+    // Until then, we just create a block under all existing blocks on the
+    // main workspace.
+
+    createAndCenterNewDefBlock(behaviorDefinitionBlock);
+  };
+  if (useModalFunctionEditor) {
+    workspace.registerButtonCallback('createNewBehavior', createNewBehavior);
+    blockList.push(newBehaviorButton);
+  } else {
+    blockList.push(behaviorDefinitionBlock);
+  }
+  blockList.push(...getCustomCategoryBlocksForFlyout('Behavior'));
+
+  const allWorkspaces = Blockly.Workspace.getAll().filter(
+    workspace => !workspace.isFlyout
+  );
+  const allBehaviorNames = [];
+  allWorkspaces.forEach(workspace => {
+    const behaviorBlocks = workspace
+      .getTopBlocks()
+      .filter(topBlock => topBlock.type === 'behavior_definition');
+    behaviorBlocks.forEach(block =>
+      allBehaviorNames.push(block.getFieldValue('NAME'))
+    );
+  });
+  allBehaviorNames.sort().forEach(name => {
+    blockList.push({
+      kind: 'block',
+      type: 'gamelab_behavior_get',
+      extraState: {
+        name: name,
+      },
+      fields: {
+        NAME: name,
+      },
+    });
+  });
+
+  return blockList;
+}
+
+function getCustomCategoryBlocksForFlyout(category) {
+  const parser = new DOMParser();
+  const xmlDoc = parser.parseFromString(Blockly.toolboxBlocks, 'text/xml');
```

Comment:
is this always xml because we aren't updating level definitions to use json? Should we put a todo here to change this when we update to using json on the levelbuilder side?

Reply:
That is my best guess. I filed a ticket for the levelbuilder work and added a comment, but I'm trying to articulate the rationale for why this would always be using XML. We can have JSON-based behaviors (user-defined, etc.), but maybe since this function starts with the pre-defined "toolbox blocks" (from levelbuilder?) and then is creating simplified blocks to add to the flyout from custom categories, we know it's XML at this point?

I think `toolboxBlocks` comes from `appOptions`: https://github.com/code-dot-org/code-dot-org/blob/staging/apps/src/code-studio/appOptions.js#L154
But Music Lab is definitely making new ones in JSON (which doesn't apply here, I don't think?): https://github.com/code-dot-org/code-dot-org/blob/staging/apps/src/music/blockly/toolbox.js#L20

Reply:
if the toolbox blocks are defined via levelbuilder, they will (for now) always be in xml. Music Lab is an exception because they (engineers) define their blocks in json. I don't think Music Lab has any intention of using behaviors.